### ISSUE TITLE
False positive `bind_instead_of_map` when panicking.

### ISSUE THREAD

**Issue by rumpuslabs** [rust-lang/rust-clippy#16861]
### Summary

Clippy thinks `Result.or_else(|_| panic!(...))` matches `Result.or_else(|x| Err(y))`.

### Lint Name

bind_instead_of_map

### Reproducer

I tried this code:

```rust
res.or_else(|_| panic!("should not happen"))
```

I saw this happen:

```
warning: using `Result.or_else(|x| Err(y))`, which is more succinctly expressed as `map_err(|x| y)`                                                                                                                       
   --> src/mod.rs:272:17
    |
272 |                 res.or_else(|_| panic!("should not happen"))
    |                 ^^^^-------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                     |
    |                     help: use `map_err` instead: `map_err`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#bind_instead_of_map
    = note: `#[warn(clippy::bind_instead_of_map)]` on by default
```

`map_err` is not incorrect, but `or_else` expresses the intent better (there's no mapping going on), and the closure does not actually contain `Err(_)` like the diagnostic says.

### Version

```text
rustc 1.93.1 (01f6ddf75 2026-02-11)
binary: rustc
commit-hash: 01f6ddf7588f42ae2d7eb0a2f21d44e8e96674cf
commit-date: 2026-02-11
host: x86_64-pc-windows-gnu
release: 1.93.1
LLVM version: 21.1.8
```
